pkg/attributes/attributes.go

```diff
@@ -0,0 +1,448 @@
+package attributes
+
+import (
+	"encoding/json"
+	"errors"
+	"strconv"
+
+	apiext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+)
+
+// Attributes provides a way to add a map of arbitrary YAML/JSON
+// objects.
+// +kubebuilder:validation:Type=object
+// +kubebuilder:validation:XPreserveUnknownFields
+type Attributes map[string]apiext.JSON
+
+// MarshalJSON implements custom JSON marshaling
+// to support free-form attributes
+func (attributes Attributes) MarshalJSON() ([]byte, error) {
+	return json.Marshal(map[string]apiext.JSON(attributes))
+}
+
+// UnmarshalJSON implements custom JSON unmarshaling
+// to support free-form attributes
+func (attributes *Attributes) UnmarshalJSON(data []byte) error {
+	return json.Unmarshal(data, (*map[string]apiext.JSON)(attributes))
+}
+
+// Exists returns `true` if the attribute with the given key
+// exists in the attributes map.
+func (attributes Attributes) Exists(key string) bool {
+	_, exists := attributes[key]
+	return exists
+}
+
+type convertPrimitiveFunc func(attributes Attributes, key string, attributeType string) (interface{}, error)
+
+func (attributes Attributes) getPrimitive(key string, zeroValue interface{}, resultType string, convert convertPrimitiveFunc, errorHolder ...*error) interface{} {
```

Comment:
Couldn't we derive `resultType` from `zeroValue`? Would it be possible to implement this function as
```go
func (attributes Attributes) getPrimitive(key string, into interface{}, convert convertPrimitiveFunc, errorHolder ...*error) {
//....
}
```
so that it could e.g. be called
```go
var value string
attributes.getPrimitive(myKey, &value, ...)
```
switching based on a raw string describing the type feels very similar to just switching on a `reflect.Type`

Reply:
well, I somehow tried, but since I didn't want to use reflection, and wanted to keep it simple (it's just an internal private helper used in 3 functions), I chose this design.

Reply:
Fair enough.